### PR TITLE
Audit implementation: correctness, streaming honesty, speed, and a smaller core

### PR DESCRIPTION
**Pull request by vboussot**
Implementation of the 2026-09-02 full-framework audit: all 122 findings processed (105 fixed, 12 dispositioned upstream with issues/PRs now filed, 4 retained with rationale, 1 rejected on evidence). Squash-merge, then tag v1.8.3 (the changelog section is included).

## Correctness
- Config binder refuses wrong-shaped YAML with the dotted path (`optimizer: AdamW` used to train without ever stepping); explicit `null` binds None; `konfai <CMD> --init` replaces the dead generation modes.
- RESUME never re-splits: the drawn seed is recorded in the workspace and read back. Exit checkpoints are honest (no junk per OOM restart, `crash_*.pt` named, BEST end no longer writes-then-deletes multi-GB).
- Nested networks dedup by identity (a same-named distinct sibling silently never trained); head resize on load is opt-in (`Model.allow_head_resize`); `Warmup` instantiates again on torch 2.12.
- Streaming budget holds its upper-bound promise on multi-Save and Expand chains (segment-keyed `SegmentSizer`); the reduction's OOM probe covers the stat pass.
- CutOUT works through YAML (fraction), LPIPS scores the whole batch, FocalLoss default is uniform, cross-group grid disagreements refuse at prepare().

## Performance
- No mid-forward host syncs in criteria (deferred one-batch readout); residual Add is a fused fold (2.3x, no N-tensor transient); SDPA attention; one-pass collate is a view; DDP one-pass shards balanced by load; DICOM streams with a budget-capped plane cache; remote OME-Zarr drops per-region round trips. Tracked `benchmarks/` harness reproduces the headline claims in one command.

## Simplification
- KonfAIInference lives in konfai-apps (loader shim keeps published bundles working); lxml, requests and huggingface_hub leave the hard dependencies; 176 private re-exports deleted with a guard test; one dev-dependency list, one mypy config, py.typed shipped; `state_dict()` is torch-native (aggregate = `network_states()`), reopening TorchScript-adjacent tooling.

## Features
- `Model.pretrained_from` (start a TRAIN from MONAI/torchvision/nnU-Net weights via the execution-order bridge), `konfai list <kind>` + `api.list_components`, prediction/evaluation per-case resume, konfai-mcp `fine_tune_app` proven by a real end-to-end test, Studio maps fine-tune jobs into its panels.

## Tests, CI, docs
- Suites run under xdist everywhere with pip caching and CPU torch on Linux; the five published `apps/*` bundle suites finally run in CI; publish is recoverable (`skip-existing`); the studio wheel's front is verified. Docs build at `-W` in CI, the PyPI page renders again, extension guides cover backends and reductions, `llms.txt` ships, and the changelog carries the full migration notes (state_dict, cutout_size, deleted criteria, modes).

Full ledger and per-area reports: `.audit-local/` (local). Watch item for this PR's CI: first xdist run on the Windows/macOS cells.

## Summary by CodeRabbit

* **New Features**
  * Added `konfai list` for component discovery and `--init` for configuration generation.
  * Added pretrained-checkpoint transfer, optional model-head resizing, resumable evaluation, prediction resume/overwrite support, and expanded streaming for imaging data.
  * Added configurable fine-tuning workflows for published apps.
  * Added reproducible benchmark tools and documentation.

* **Bug Fixes**
  * Improved configuration validation, checkpoint handling, dataset caching, streaming consistency, and TensorBoard fallback behavior.
  * Improved DICOM and remote dataset read performance.

* **Documentation**
  * Expanded guidance for CLI usage, workflows, extension points, metrics, models, and benchmarks.
  * Documented updated checkpoint, configuration, selection, and fine-tuning workflows.

### CODE REVIEW

File: benchmarks/bench_streaming.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    return total
+
+
+class PeakSampler(threading.Thread):
+    """Whole-tree peak RSS, sampled at 50 ms: children (DataLoader workers, spawn ranks) count."""
+
+    def __init__(self) -> None:
+        super().__init__(daemon=True)
+        self._process = psutil.Process()
+        self._stop = threading.Event()
```
[the remark this answers is not in the document]
Comment (vboussot) [reply benchmarks/bench_streaming.py]: Fixed: renamed to self._stop_event (reproduced the TypeError from Thread.join calling the internal _stop()).

File: benchmarks/bench_streaming.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+def main() -> None:
+    parser = argparse.ArgumentParser(description=__doc__)
+    parser.add_argument("--gib", type=float, default=4.0, help="synthetic volume size in GiB (default 4)")
+    parser.add_argument("--budget", type=float, default=1.0, help="declared memory_budget in GiB (default 1)")
+    parser.add_argument("--keep", action="store_true", help="keep the scratch directory")
+    args = parser.parse_args()
+
+    import konfai
+
+    scratch = Path(tempfile.mkdtemp(prefix="konfai_bench_"))
```
[the remark this answers is not in the document]
Comment (vboussot) [reply benchmarks/bench_streaming.py:101]: Fixed: the benchmark body now runs in try/finally and the finalizer removes the scratch directory unless --keep is passed.

File: docs/source/usage/benchmarks.md [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+the same footing.
+
+## Protocol
+
+- Wall time is the median of 3 runs after 1 warmup, on an otherwise idle
+  machine.
+- Host memory is the peak resident set of the whole process tree (`psutil`),
+  sampled at 50 ms, so DataLoader workers and spawned ranks count.
+- Device memory is `torch.cuda.max_memory_allocated()` plus the NVML process
+  figure when available.
```
[the remark this answers is not in the document]
Comment (vboussot) [reply docs/source/usage/benchmarks.md]: Fixed: the protocol line now says the NVML per-process figure is reported beside torch.cuda.max_memory_allocated(), and that the two overlap and are never summed.

File: docs/source/usage/benchmarks.md [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+- Every report line carries the konfai/torch/SimpleITK versions, CPU model,
+  GPU model, and the input's shape, dtype and checksum.
+
+## The streaming claim
+
+{doc}`../concepts/streaming` states that a case never has to fit in RAM: a
+16 GiB uncompressed volume is transformed with peak host memory bounded by the
+declared `memory_budget`, not by the volume. Reproduce it with one command from
```
[the remark this answers is not in the document]
Comment (vboussot) [reply docs/source/usage/benchmarks.md]: Fixed: the claim is now scoped to a streamable case, with a sentence stating that a non-streamable chain falls back to the whole volume, which TRANSFORM sizes against the same budget and refuses when it does not fit (a designed refusal, not an OOM: the plan prices the whole-volume route before a byte is read).

File: konfai/data/augmentation/spatial.py [hunk 5]
```diff
@@ -546,88 +508,142 @@ def _inverse(self, index: int, a: int, tensor: torch.Tensor) -> torch.Tensor:
...
+        target_slices: tuple[slice, ...],
+        source_spatial_shape: list[int],
+    ) -> list[slice]:
+        # |displacement| <= max_displacement in world units by convexity: in voxels that is the
+        # bound over the axis's spacing, plus one voxel for the far interpolation tap.
+        _stage, grid = self.draws[index][a]
+        rank = len(source_spatial_shape)
+        pull: list[slice] = []
+        for k, (part, extent) in enumerate(zip(target_slices, source_spatial_shape, strict=True)):
+            reach = int(np.ceil(self.max_displacement / float(grid.spacing_xyz[rank - 1 - k]))) + 1
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai/data/augmentation/spatial.py]: Fixed: the reach per index axis is now max_displacement times the L1 norm of that axis's row of grid.world_to_index.matrix (reduces to 1/spacing on an axis-aligned grid). Added test_elastix_streams_each_region_of_an_oblique_copy with a 45-degree Direction and a worst-case corner draw; it shows a 0.63 max abs error under the old spacing-only reach and 1e-6 under the row-norm reach.

File: konfai/data/geometry.py [hunk 6]
```diff
@@ -371,6 +371,96 @@ def sub_grid(self, region_zyx: tuple[slice, ...]) -> Grid:
...
+    carries unit weight. The three tests together admit exactly those: unit column sums alone also
+    pass an axis-averaging matrix, unit peaks alone a superposing one, and columns alone a
+    rank-deficient one that reads the same axis twice. The answer is in array order, where physical
+    axis ``k`` is array axis ``n - 1 - k``.
+    """
+    linear = np.asarray(matrix, dtype=np.float64)
+    n = int(linear.shape[0])
+    magnitude = np.abs(linear)
+    unit = np.ones(n)
+    if not np.allclose(magnitude.sum(axis=0), unit, atol=atol):
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai/data/geometry.py]: Fixed: rtol=0.0 on all three np.allclose calls, so only the declared atol constants control acceptance. Added test_a_near_axis_rotation_is_refused (theta=5e-6, which passed under the default rtol).

File: konfai/data/patching/manager.py [hunk 7]
```diff
@@ -1569,63 +1337,60 @@ def working_multiple(self) -> float:
...
+            return self.read_granularity()
+        key = (str(dataset.filename), group, entry)
+        if key not in self._granularities:
+            granularity = None
+            # A cache this run has still to write has no metadata to ask; its chunks will be the
+            # very tile being sized, so its reads align by construction and None is its honest grain.
+            if dataset.is_dataset_exist(group, entry):
+                stored = dataset.read_granularity(group, entry)
+                granularity = None if stored is None else tuple(stored[1:])
+            self._granularities[key] = granularity
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai/data/patching/manager.py]: Fixed: a missing entry now returns None without being memoized, so a downstream segment re-planned after _invalidate_stream_plans() (which never cleared _granularities) prices its reads on the published cache's stored chunks instead of a stale exact-read grain.

File: konfai/data/transform/intensity.py [hunk 8]
```diff
@@ -282,41 +363,79 @@ def __init__(
...
+            raise TransformError(
+                f"The mask '{self.mask}' has shape {list(mask.shape)} where the tensor in hand has"
+                f" {list(tensor.shape)}: it cannot be indexed against a region.",
+                "A masked statistic needs the whole volume here; report this if the chain was planned.",
+            )
+        return tensor[mask == 1]
+
     def __call__(self, name: str, tensor: torch.Tensor, cache_attribute: Attribute) -> torch.Tensor:
-        tensor_masked = tensor if self.mask is None else tensor[self.read_companion(self.mask, name) == 1]
+        if self.mask is not None and "StatisticsSeeded" in cache_attribute:
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai/data/transform/intensity.py]: Fixed: the seeded branch is now gated on self.mean is None or self.std is None, matching patch_locality's POINTWISE declaration for fully configured coefficients; the else branch never calls _masked_values when both are given, so no mask is read.

File: konfai/network/network/measure.py [hunk 9]
```diff
@@ -32,6 +33,73 @@
...
+        if isinstance(value, np.generic):
+            value = float(value)
+        elif isinstance(value, bool | int):
+            value = float(value)
+        elif isinstance(value, tuple) and not isinstance(value, LabelledValues):
+            if len(value) == 2 and isinstance(value[0], torch.Tensor) and isinstance(value[1], list):
+                value = LabelledValues(value[0], value[1])
+        if not isinstance(value, float | torch.Tensor | dict | LabelledValues) or (
+            map_ is not None and not isinstance(map_, torch.Tensor)
+        ):
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai/network/network/measure.py:99]: Fixed: CriterionResult.of now raises MeasureError for a plain tensor value with numel() != 1, and the bare-tensor return path funnels through the tuple path so a reduction='none' loss is caught by the same check instead of silently shifting the per-device readout iterator.

File: konfai/network/network/network.py [hunk 10]
```diff
@@ -572,21 +593,37 @@ def __init__(
...
+        if destination is None:
+            destination = OrderedDict()
         local_metadata = {"version": self._version}
-        self._save_to_state_dict(destination, "", False)
+        self._save_to_state_dict(destination, prefix, keep_vars)
         for name, module in self._modules.items():
             if module is not None:
                 if not isinstance(module, Network):
-                    module.state_dict(destination=destination, prefix="" + name + ".", keep_vars=False)
+                    module.state_dict(destination=destination, prefix=prefix + name + ".", keep_vars=keep_vars)
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai/network/network/network.py:628]: Fixed: state_dict now initialises destination._metadata and records {'version': self._version} per prefix (non-Network children then record theirs through the torch recursion), and Network.load carries the source _metadata onto the rebuilt model_state_dict with its module paths remapped through the same alias rule as the tensors. Also dropped the state_dict['_metadata'] = metadata KEY insertion in load_state_dict, which the strict per-module pass would have collected as an unexpected key once metadata became non-None.

File: konfai/network/network/network.py [hunk 11]
```diff
@@ -714,19 +756,21 @@ def load(
...
+                # Segment-aligned: alias 'layer1' must not claim a module 'layer10', and only the
+                # leading prefix is rewritten, never a later occurrence of the same substring.
+                alias_list = [(a, b) for a, b in modules_name.items() if prefix == a or prefix.startswith(a + ".")]
 
                 if len(alias_list):
-                    for a, b in alias_list:
-                        model_state_dict[alias.replace(a, b)] = model_state_dict_tmp[alias]
-                        break
+                    a, b = alias_list[0]
+                    model_state_dict[b + alias[len(a) :]] = model_state_dict_tmp[alias]
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai/network/network/network.py]: Fixed: the matching candidates now resolve through a remap helper that takes the longest matching alias (max by len), so a nested alias like p.a1 wins over its parent p regardless of insertion order; the same helper remaps the load metadata paths.

File: konfai/trainer.py [hunk 12]
```diff
@@ -1222,7 +1232,7 @@ def train(
     command: State = State.TRAIN,
     overwrite: bool = False,
     model: Path | str | None = None,
-    gpu: list[int] | None = cuda_visible_devices(),
+    gpu: list[int] | None = None,
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai/trainer.py:1235]: Intentional, not changed: gpu=None means CPU at every entrypoint by design (this PR's audit decision: no CUDA probe when resolving defaults, devices are selected explicitly). The CLI documents it ('--gpu ... If omitted runs on CPU'), the launcher prints one startup line naming the resolved devices, and auto-detecting visible GPUs would reintroduce the import-time CUDA probe the change removed.

File: konfai/utils/config.py [hunk 13]
```diff
@@ -458,8 +413,20 @@ def decorator(function):
...
+
+def _tensor_type() -> type | None:
+    """``torch.Tensor`` when torch is already imported, else None.
+
+    torch is never imported here: an annotation can only mention Tensor if its declaring module
+    already paid the import, and keeping config.py torch-free keeps the light-import contract of
+    ``konfai/__init__`` honest for consumers like the Slicer-facing konfai-apps helpers (measured:
+    634 of this module's 676 ms import was torch).
+    """
+    return getattr(sys.modules.get("torch"), "Tensor", None)
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai/utils/config.py:427]: Fixed: _bind_parameter now dispatches a bare torch.Tensor annotation (which Optional[Tensor] also unwraps to) through _convert_union_sequence_value, so the configured value binds as a tensor instead of falling into _bind_config_object. Regression test added: test_apply_config_binds_a_bare_tensor_parameter_as_a_value.

File: benchmarks/bench_streaming.py [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+                "numpy": np.__version__,
+                "python": platform.python_version(),
+            },
+            "host": platform.node(),
+        }
+        print(json.dumps(report, indent=2))
+        print(
+            f"| {args.gib:g} GiB volume | budget {args.budget:g} GiB | peak {peak / 2**30:.2f} GiB | {elapsed:.1f} s |"
+        )
+    finally:
```
[the remark this answers is not in the document]
Comment (vboussot) [reply benchmarks/bench_streaming.py:142]: Fixed in 43d1625a.

File: konfai/data/patching/manager.py [hunk 15]
```diff
@@ -1569,63 +1337,60 @@ def working_multiple(self) -> float:
...
+            return self.read_granularity()
+        key = (str(dataset.filename), group, entry)
+        if key not in self._granularities:
+            # A cache this run has still to write has no metadata to ask. That answer is not
+            # memoized: once the upstream sweep publishes the cache, a re-planned downstream
+            # segment must price its reads on the stored chunks, not on a stale exact-read grain.
+            if not dataset.is_dataset_exist(group, entry):
+                return None
+            stored = dataset.read_granularity(group, entry)
+            self._granularities[key] = None if stored is None else tuple(stored[1:])
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai/data/patching/manager.py:1392]: Fixed in 43d1625a.

File: konfai/network/network/measure.py [hunk 16]
```diff
@@ -32,6 +33,81 @@
...
+        elif isinstance(value, tuple) and not isinstance(value, LabelledValues):
+            if len(value) == 2 and isinstance(value[0], torch.Tensor) and isinstance(value[1], list):
+                value = LabelledValues(value[0], value[1])
+        if isinstance(value, torch.Tensor) and value.numel() != 1:
+            # A multi-element plain value would silently shift the deferred per-device readout.
+            raise MeasureError(
+                f"'{criterion}' reported a tensor of {value.numel()} elements.",
+                "The reported value is a single number: reduce it, or report a (values, labels) "
+                "pair for a per-label metric.",
+            )
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai/network/network/measure.py:90]: Fixed in 43d1625a.
